File: Student_Unable_Eat_Lunch/count_students.dart
class Node {
  int val;
  Node? next;
  Node(this.val, this.next);
}

class Queue {
  Node? head;
  Node? tail;
  int size;
  Queue(this.head, this.tail, this.size);

  bool IsEmpty() {
    return size == 0;
  }

  void Enqueue(int val) {
    if (IsEmpty()) {
      Node node = Node(val, null);
      head = node;
      tail = node;
    } else {
      Node node = Node(val, null);
      tail?.next = node;
      tail = node;
    }
    size++;
  }

  Node? Dequeue() {
    if (IsEmpty()) {
      throw Exception("Queue is empty, cannot dequeue");
    }
    Node removed = head!;
    head = head!.next;

    if (head == null) {
      tail = null;
    }
    size--;
    return removed;
  }

  int Front() {
    if (IsEmpty()) {
      throw Exception("Cannot return value of an empty queue");
    }

    return head!.val;
  }
}

class Solution {
  int countStudents(List<int> students, List<int> sandwiches) {
    Queue studentQueue = Queue(null, null, students.length);

    for (var student in students) {
      studentQueue.Enqueue(student);
    }

    for (var sandwich in sandwiches) {
      int maxAttempt = students.length;

      while (maxAttempt > 0 && studentQueue.Front() != sandwich) {
        Node? removed = studentQueue.Dequeue();

        if (removed != null) {
          studentQueue.Enqueue(removed.val);
        }
        maxAttempt--;
      }

      if (maxAttempt == 0) {
        break;
      }

      studentQueue.Dequeue();
    }

    return studentQueue.size;
  }
}

File: Student_Unable_Eat_Lunch/count_students.go
package studentunableeatlunch

type Node struct {
	val  int
	next *Node
}

type Queue struct {
	head *Node
	tail *Node
	size int
}

func (q *Queue) IsEmpty() bool {
	return q.size == 0
}

func (q *Queue) Enqueue(val int) {
	node := &Node{val: val, next: nil}

	if q.IsEmpty() {
		q.head = node
		q.tail = node
	} else {
		q.tail.next = node
		q.tail = node
	}
	q.size++
}

func (q *Queue) Dequeue() *Node {
	if q.IsEmpty() {
		return nil
	}

	dequeueNode := q.head
	q.head = q.head.next

	if q.head == nil {
		q.tail = nil
	}

	q.size--
	return dequeueNode
}

func (q *Queue) Front() int {
	if q.IsEmpty() {
		return -1
	}

	return q.head.val
}

func countStudents(students []int, sandwiches []int) int {
	studentQueue := &Queue{}

	for _, v := range students {
		studentQueue.Enqueue(v)
	}

	for _, v := range sandwiches {
		attemps := studentQueue.size
		for attemps > 0 && studentQueue.Front() != v {
			removed := studentQueue.Dequeue()
			studentQueue.Enqueue(removed.val)
			attemps--
		}

		if attemps == 0 {
			break
		}

		studentQueue.Dequeue()
	}

	return studentQueue.size
}
